File: packages/react-native-worklets/android/src/main/java/com/swmansion/worklets/WorkletsPackage.java
package com.swmansion.worklets;

import androidx.annotation.NonNull;
import com.facebook.react.BaseReactPackage;
import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.module.annotations.ReactModuleList;
import com.facebook.react.module.model.ReactModuleInfo;
import com.facebook.react.module.model.ReactModuleInfoProvider;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@ReactModuleList(nativeModules = {WorkletsModule.class})
public class WorkletsPackage extends BaseReactPackage implements ReactPackage {
  @Override
  public NativeModule getModule(
      @NonNull String name, @NonNull ReactApplicationContext reactContext) {
    return name.equals(WorkletsModule.NAME) ? new WorkletsModule(reactContext) : null;
  }

  @SuppressWarnings({"rawtypes, unchecked"})
  @NonNull
  @Override
  public ReactModuleInfoProvider getReactModuleInfoProvider() {
    Class[] moduleList = new Class[] {WorkletsModule.class};

    final Map<String, ReactModuleInfo> reactModuleInfoMap = new HashMap<>();
    for (Class<? extends NativeModule> moduleClass : moduleList) {
      ReactModule reactModule =
          Objects.requireNonNull(moduleClass.getAnnotation(ReactModule.class));

      reactModuleInfoMap.put(
          reactModule.name(),
          new ReactModuleInfo(
              reactModule.name(),
              moduleClass.getName(),
              reactModule.canOverrideExistingModule(),
              reactModule.needsEagerInit(),
              reactModule.isCxxModule(),
              true));
    }

    return () -> reactModuleInfoMap;
  }
}

File: packages/react-native-worklets/android/src/main/java/com/swmansion/worklets/WorkletsPackage.kt
package com.swmansion.worklets

import com.facebook.react.BaseReactPackage
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.module.annotations.ReactModuleList
import com.facebook.react.module.model.ReactModuleInfo
import com.facebook.react.module.model.ReactModuleInfoProvider

@ReactModuleList(nativeModules = [WorkletsModule::class])
class WorkletsPackage : BaseReactPackage(), ReactPackage {

    override fun getModule(name: String, reactContext: ReactApplicationContext): NativeModule? =
        if (name == WorkletsModule.NAME) WorkletsModule(reactContext) else null

    override fun getReactModuleInfoProvider(): ReactModuleInfoProvider {
        val moduleList: Array<Class<*>> = arrayOf(WorkletsModule::class.java)

        val reactModuleInfoMap = HashMap<String, ReactModuleInfo>()
        for (moduleClass in moduleList) {
            val reactModule: ReactModule =
                checkNotNull(moduleClass.getAnnotation(ReactModule::class.java))

            reactModuleInfoMap[reactModule.name] =
                ReactModuleInfo(
                    reactModule.name,
                    moduleClass.name,
                    reactModule.canOverrideExistingModule,
                    reactModule.needsEagerInit,
                    reactModule.isCxxModule,
                    true
                )
        }

        return ReactModuleInfoProvider { reactModuleInfoMap }
    }
}
